### DISCUSSION TITLE
What is the difference and purpose of simulation and emulation

### DISCUSSION THREAD

**Discussion by IAmMarcelJung** [FPGA-Research/FABulous#219] [Q&A]
As this came up in #218, here the difference and purpose of both the simulation and emulation in FABulous can be discussed.

In my understanding, the simulation would run the fabric inside a testbench, including the uploading of the user design bitstream  (which `run_simulation.sh` currently does) and the emulation would run the fabric on acutal FPGA hardware (which `run_emulation_sim.sh` currently does **not** do, as far as I understand.) 

What is certain is that the [documentation of the emulation](https://fabulous.readthedocs.io/en/latest/simulation/emulation.html) needs to be updated since there are some flags mentioned which do not exist anymore or at least renamed.

**Comment by KelvinChung2000** [comment]
I have taken a quick look at the testbench and the documentation. It looks like the emulation skips the bitstream configuration part and wires the bitstream value into the fabric for faster simulation time. So, what we have in "simulation" will be emulation by my definition.  

I would say running the design on hardware can also involve simulation and emulation. For example, building the fabric with the bitstream preconfigured and downloading it to the board will be a hardware simulation. Building the fabric, loading the fabric onto the hardware, and then putting the test bitstream onto the device with the configuration mechanism will be hardware emulation.

**Reply by IAmMarcelJung** [reply]
Thanks for looking into this and your thought on this! After looking through the answers on a [SO post](https://stackoverflow.com/questions/1584617/simulator-or-emulator-what-is-the-difference) and thinking about them, my personal understanding is this: Simulation is used to find out about the inner workings of design and also test them (e.g. SPICE simlation). Emulation does not really care about the internals and just creates something that mimics the behaviour of the thing to be emulated (e.g. a game console emulator).

In the end, we don't really have care what the "official" definiton for the terms is, since the terms are also used interchangebly in different contexts, but we should define what simulation and emulation does in FABulous, what the reasoning for both is and name the scripts correspondingly.

**Comment by EverythingElseWasAlreadyTaken** [comment]
It seems that the bash scrips are kind of deprecated anyway, they still use the old synth_fabulous.tcl, prims.v etc., which is already included in upstream yosys in a much better way: https://github.com/YosysHQ/yosys/tree/main/techlibs/fabulous.

The run_FABulous_bitstream command already only uses the yosys built-in version, so we actually don't have to ship these files anymore with FABulous demo projects.
